### PR TITLE
chore: change survey prompt library and add test with it

### LINKED ISSUE #36: Test survey prompts
In our unit tests, we currently shortcut all survey prompts creating code paths that are not covered. We should find a way to adequately provide input to the prompts in order to cover those cases.

issue originally submitted by @scothis

### PR DESCRIPTION
**Pull request by odinnordico**
Pull request

### What this PR does / why we need it

Changed the used library for survey prompt to use a simple one

### Which issue(s) this PR fixes

Fixes #36 

### Describe testing done for PR

Unit test running in MacOS and Linux

### Additional information or special notes for your reviewer

Survey prompts confirmation options are changing from `(y/N)` to  `[yN]` but the behavior is the same

### CODE REVIEW

**Review by warango4 — COMMENTED** [review]
This is really a great job

File: docs/commands-details/workload_create_update_apply.md [hunk 1]
```diff
@@ -88,7 +88,7 @@ Update workload:
 18, 17   |        tag: tap-1.1
 ...
 
-? Really update the workload "spring-pet-clinic"? (y/N)
+? Really update the workload "spring-pet-clinic"? [yN]
```
Comment (warango4) [inline docs/commands-details/workload_create_update_apply.md:91]: I would be inclined to ask design if it's correct to use the survey this way (`[yN]`)

Comment (odinnordico) [reply docs/commands-details/workload_create_update_apply.md:91]: @danfein please a check here

File: pkg/commands/workload_create_test.go [hunk 2]
```diff
@@ -1296,6 +1297,95 @@ To get status: "tanzu apps workload get my-workload"
...
+      9 + |spec:
+     10 + |  source:
+     11 + |    git:
+     12 + |      ref:
+     13 + |        branch: main
+     14 + |      url: https://example.com/repo.git
+
+%s
+
+Skipping workload %q`, clitesting.ToInteractTerminal("? Do you want to create this workload? [yN]: "), workloadName),
```
Comment (shaheerkootteeri) [inline pkg/commands/workload_create_test.go:1387]: test case not capturing the response `n` in the expected output

File: pkg/commands/workload_apply_test.go [hunk 3]
```diff
@@ -3279,6 +3280,220 @@ To get status: "tanzu apps workload get my-workload"
...
+      6 + |    apps.tanzu.vmware.com/workload-type: api
+  7,  7   |  name: my-workload
+  8,  8   |  namespace: default
+  9,  9   |spec:
+ 10, 10   |  source:
+...
+
+%s
+
+Skipping workload %q`, clitesting.ToInteractTerminal("? Really update the workload %q? [yN]: ", workloadName), workloadName),
```
Comment (shaheerkootteeri) [inline pkg/commands/workload_apply_test.go:3495]: test case not capturing the response `n` in the expected output

File: testing/e2e/workload_test.go [hunk 4]
```diff
@@ -65,8 +65,8 @@ func TestCreateFromGitWithAnnotations(t *testing.T) {
 					"--git-repo=https://github.com/sample-accelerators/spring-petclinic",
 					namespaceFlag,
 					"--type=web",
-					"--yes",
```
Comment (shaheerkootteeri) [inline testing/e2e/workload_test.go:68]: shouldn't `--yes` skip the survey prompt ?

**Review by shaheerkootteeri — CHANGES_REQUESTED** [review]

File: cmd/plugin/apps/main.go [hunk 5]
```diff
@@ -98,9 +97,6 @@ func main() {
 	p.Cmd.PersistentFlags().Int32VarP(c.Verbose, cli.StripDash(flags.VerboseLevelFlagName), "v", 1, "number for the log level verbosity")
 
 	cobra.OnInitialize(func() {
-		// sync survey and faith option to disable color
```
[the remark this answers is not in the document]
Comment (shashwathi) [reply cmd/plugin/apps/main.go:101]: +1

File: pkg/cli-runtime/survey.go [hunk 6]
```diff
@@ -0,0 +1,17 @@
+package cli
+
+import (
+	"fmt"
+
+	"github.com/vito/go-interact/interact"
+
```
[the remark this answers is not in the document]
Comment (shashwathi) [reply pkg/cli-runtime/survey.go:7]: The imprts are grouped in 3 groups. 
1st group : base golang dependencies
2nd group: external libraries
3rd group: our own lib deps. 
This follows that rule so resolving it

File: pkg/cli-runtime/testing/command_table.go [hunk 7]
```diff
@@ -304,3 +343,60 @@ func fakeExecCommand(helper string) func(context.Context, string, ...string) *ex
 		return cmd
 	}
 }
+
+func trimTrailingSpaces(in string) string {
+	lines := strings.Split(in, "\n")
+	for i := len(lines) - 1; i >= 0; i-- {
+		lines[i] = strings.TrimRight(lines[i], " ")
+	}
+	return strings.Join(lines, "\n")
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/command_table.go:352]: declare line break constant in separate lib and use that across all files. When the time comes to add support for line breaks across diff platforms then we can make that change easily.

File: pkg/cli-runtime/survey.go [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+	"github.com/vmware-tanzu/apps-cli-plugin/pkg/cli-runtime/printer"
+)
+
+// NewConfirmSurvey create a survey asking for [yN] confirmation when `Resolve` is called
+func NewConfirmSurvey(c *Config, format string, a ...any) *interact.Interaction {
+	i := interact.NewInteraction(fmt.Sprintf("%s %s", printer.Ssuccessf("?"), printer.Sboldf(fmt.Sprintf(format, a...))))
```
Comment (shashwathi) [inline pkg/cli-runtime/survey.go:13]: why this interaction call vs choice func? 
ref: https://github.com/vito/go-interact/blob/master/interact/choice.go
Want to understand your reasoning. can we use choice with `y` and `N` as options?

**Review by shashwathi — COMMENTED** [review]
add test case for scenario where use enters something other than y/n so we are aware of the behaviour.

File: pkg/cli-runtime/survey.go (hunk 8 above)
Comment (odinnordico) [reply pkg/cli-runtime/survey.go:13]: Choice will show the options like:
```console
1: y
2: n
```
without case validation, `NewConfirmSurvey`, when no *options* are used defaults to bolean validation with `y` & `n` values promt with case validation

File: testing/e2e/workload_test.go (hunk 4 above)
Comment (odinnordico) [reply testing/e2e/workload_test.go:68]: yes but the survey prompt is being managed by `c.SurveyAnswer("y")` for this test

File: pkg/cli-runtime/testing/command_table.go [hunk 9]
```diff
@@ -251,26 +257,58 @@ func (tc CommandTestCase) Run(t *testing.T, scheme *runtime.Scheme, cmdFactory f
...
+		cmd.SetOutput(c.Stdout)
+
 		cmdErr := cmd.Execute()
 
+		if console != nil {
+			console.Tty().Close()
+		}
+
+		if donec != nil {
+			<-donec
```
[the remark this answers is not in the document]
Comment (odinnordico) [reply pkg/cli-runtime/testing/command_table.go:310]: The channel is used only for sync, no value nor error is send

File: cmd/plugin/apps/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (odinnordico) [reply cmd/plugin/apps/main.go:101]: the new prompt lib does not have color management, it is managed with plugin color lib

File: pkg/cli-runtime/testing/command_table.go [hunk 10]
```diff
@@ -304,3 +343,60 @@ func fakeExecCommand(helper string) func(context.Context, string, ...string) *ex
...
+
+func trimTrailingSpaces(in string) string {
+	lines := strings.Split(in, "\n")
+	for i := len(lines) - 1; i >= 0; i-- {
+		lines[i] = strings.TrimRight(lines[i], " ")
+	}
+	return strings.Join(lines, "\n")
+}
+
+func expectNoMatchingError(t *testing.T) expect.ConsoleOpt {
```
[the remark this answers is not in the document]
Comment (odinnordico) [reply pkg/cli-runtime/testing/command_table.go]: not required, removed

File: pkg/cli-runtime/testing/command_table.go [hunk 11]
```diff
@@ -304,3 +343,60 @@ func fakeExecCommand(helper string) func(context.Context, string, ...string) *ex
...
+				for _, matcher := range matchers {
+					criteria = append(criteria, fmt.Sprintf("%q", matcher.Criteria()))
+				}
+				t.Fatalf("Unexpected output; expected: %s ; got %q: \nError: %s\n", strings.Join(criteria, ", "), buf, err)
+			}
+		},
+	)
+}
+
+func expectNoInputError(t *testing.T) expect.ConsoleOpt {
```
[the remark this answers is not in the document]
Comment (odinnordico) [reply pkg/cli-runtime/testing/command_table.go]: not required, removed

File: pkg/commands/workload_create_test.go (hunk 2 above)
Comment (odinnordico) [reply pkg/commands/workload_create_test.go]: Fixed, bad skipping char

File: pkg/commands/workload_apply_test.go (hunk 3 above)
Comment (odinnordico) [reply pkg/commands/workload_apply_test.go]: Fixed, bad skipping char

File: pkg/cli-runtime/testing/command_table.go (hunk 7 above)
Comment (odinnordico) [reply pkg/cli-runtime/testing/command_table.go]: test constants used instead

File: pkg/cli-runtime/testing/constants.go [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+const (
+	NewLine         = "\n"
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/constants.go:20]: this is useful outside of testing. i would move this into a package at `/pkg/` level

File: pkg/cli-runtime/testing/command_table.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (shashwathi) [reply pkg/cli-runtime/testing/command_table.go:310]: why is this channel required here? I am missing the purpose, console executes all expectations in go routine.

File: pkg/cli-runtime/testing/command_table.go [hunk 13]
```diff
@@ -251,26 +257,58 @@ func (tc CommandTestCase) Run(t *testing.T, scheme *runtime.Scheme, cmdFactory f
...
+
+		cmd := cmdFactory(ctx, c)
+		cmd.SilenceErrors = true
+		cmd.SilenceUsage = true
+		cmd.SetArgs(tc.Args)
+		cmd.SetOutput(c.Stdout)
+
 		cmdErr := cmd.Execute()
 
+		if console != nil {
```
Comment (shashwathi) [inline pkg/cli-runtime/testing/command_table.go:305]: why not close the console in go func? once expecations are checked, call defer on console.Close(). wdyt

File: pkg/cli-runtime/testing/command_table.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (shashwathi) [reply pkg/cli-runtime/testing/command_table.go:310]: donec is used to just mark whether console expectations were met or not.

File: cmd/plugin/apps/main.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (odinnordico) [reply cmd/plugin/apps/main.go:101]: the `--no-color` binding is `color.NoColor`, which is the one used by the `printers`, then the output will be with no color when the flag is present

**Review by shashwathi — APPROVED** [review]

**Review by warango4 — APPROVED** [review]

**Review by shaheerkootteeri — APPROVED** [review]

**Pull request merged**